barretenberg/cpp/src/barretenberg/vm/avm/trace/deserialization.cpp

```diff
@@ -186,112 +187,152 @@ const std::unordered_map<OpCode, std::vector<OperandType>> OPCODE_WIRE_FORMAT =
     { OperandType::UINT8, 1 },     { OperandType::UINT16, 2 },     { OperandType::UINT32, 4 },
     { OperandType::UINT64, 8 },    { OperandType::UINT128, 16 },   { OperandType::FF, 32 }
 };
 
 } // Anonymous namespace
 
+// TODO: once opcodes are frozen, this function can be replaced by a table/map of constants
+uint32_t Deserialization::get_pc_increment(OpCode opcode)
+{
+    const auto iter = OPCODE_WIRE_FORMAT.find(opcode);
+
+    if (iter == OPCODE_WIRE_FORMAT.end()) {
+        return 0;
+    }
+
+    // OPCODE_WIRE_FORMAT does not contain the opcode itself which accounts for 1 byte
+    uint32_t increment = 1;
+
+    const std::vector<OperandType>& inst_format = iter->second;
+    for (const auto& op_type : inst_format) {
+        increment += OPERAND_TYPE_SIZE.at(op_type);
+    }
+
+    return increment;
+}
+
 /**
- * @brief Parsing of the supplied bytecode into a vector of instructions. It essentially
- *        checks that each opcode value is in the defined range and extracts the operands
+ * @brief Parsing of an instruction in the supplied bytecode at byte position pos. This
+ *        checks that the opcode value is in the defined range and extracts the operands
  *        for each opcode based on the specification from OPCODE_WIRE_FORMAT.
  *
  * @param bytecode The bytecode to be parsed as a vector of bytes/uint8_t
- * @throws runtime_error exception when the bytecode is invalid.
- * @return Vector of instructions
+ * @param pos Bytecode position
+ * @throws runtime_error exception when the bytecode is invalid or pos is out-of-range
+ * @return The instruction
  */
-std::vector<Instruction> Deserialization::parse(std::vector<uint8_t> const& bytecode)
+Instruction Deserialization::parse(const std::vector<uint8_t>& bytecode, size_t pos)
 {
-    std::vector<Instruction> instructions;
-    size_t pos = 0;
     const auto length = bytecode.size();
 
-    debug("------- PARSING BYTECODE -------");
-    debug("Parsing bytecode of length: " + std::to_string(length));
-    while (pos < length) {
-        const uint8_t opcode_byte = bytecode.at(pos);
+    if (pos >= length) {
+        throw_or_abort("Position is out of range. Position: " + std::to_string(pos) +
+                       " Bytecode length: " + std::to_string(length));
+    }
+
+    const uint8_t opcode_byte = bytecode.at(pos);
 
-        if (!Bytecode::is_valid(opcode_byte)) {
-            throw_or_abort("Invalid opcode byte: " + to_hex(opcode_byte) + " at position: " + std::to_string(pos));
-        }
-        pos++;
+    if (!Bytecode::is_valid(opcode_byte)) {
+        throw_or_abort("Invalid opcode byte: " + to_hex(opcode_byte) + " at position: " + std::to_string(pos));
+    }
+    pos++;
+
+    const auto opcode = static_cast<OpCode>(opcode_byte);
+    const auto iter = OPCODE_WIRE_FORMAT.find(opcode);
+    if (iter == OPCODE_WIRE_FORMAT.end()) {
+        throw_or_abort("Opcode not found in OPCODE_WIRE_FORMAT: " + to_hex(opcode) + " name " + to_string(opcode));
+    }
```

Comment:
What would it mean for the wire format not to include the opcode? Is this basically a redundant check that "opcode is valid"?

Reply:
I see now that this isn't new code and that the only diff for it is that it was pulled into a helper function and indentation changed